polyhook2/Virtuals/VTableSwapHook.hpp

```diff
@@ -40,5 +41,64 @@ class VTableSwapHook : public PLH::IHook {
 	VFuncMap m_origVFuncs;
 	bool m_Hooked;
 };
+
+
+// storage class for address of a virtual function
+// also stores the function pointer type and index number on the class level
+template<uint16_t I, typename FuncPtr>
+struct VFunc {
+	VFunc() : func(nullptr) {};
+	VFunc(FuncPtr f) : func(f) {};
+	const FuncPtr func;
+	static const uint16_t func_index;
+	typedef FuncPtr func_type;
+};
+
+// definition of constant must reside outside class declaration
+template<uint16_t I, typename FuncPtr> const uint16_t VFunc<I, FuncPtr>::func_index = I;
+
+namespace detail {
+
+	// helper function to convert sequence of VFunc structs into a VFuncMap
+	// using recursive template definition
+	inline PLH::VFuncMap make_vfunc_map()
+	{
+		return PLH::VFuncMap{ };
+	};
+
+	template<uint16_t I, typename FuncPtr, typename ... VFuncTypes>
+	inline PLH::VFuncMap make_vfunc_map(VFunc<I, FuncPtr> vfunc, VFuncTypes ... vfuncs)
+	{
+		PLH::VFuncMap map{ {I, reinterpret_cast<uint64_t>(vfunc.func)} };
+		map.merge(make_vfunc_map(vfuncs ...));
+		return map;
+	};
+
+}
+
+template<class ClassType, typename ... VFuncTypes>
+class VTableSwapHook2 : private VTableSwapHook {
+public:
+	VTableSwapHook2(ClassType* instance, VFuncTypes ... newFuncs)
+		: PLH::VTableSwapHook((char*)instance, detail::make_vfunc_map(newFuncs ...))
+	{
+		if (!hook())
+			throw std::runtime_error(std::string("failed to hook ") + typeid(ClassType).name());
```

Comment:
Polyhook chooses not to use C++ exceptions due to the fact that many people manual map or otherwise use polyhook in environments that do not support exception handling. Please prefer the Init() or Hook() pattern that the rest of the library uses.

Reply:
Sure, I will keep this in mind in future patches.